File: combinationsumiii.js
//Objective is, given 'n' and 'k', which represent the number of elements in the output
//and the target value that the elements add up to, to find all unique combinations from 
//1 - 9.

let k = 3 
let n = 7


//O() solution

let temp = []
let result = []
let sum = 0
let index = 1

backtrack(temp, sum, index)

function backtrack(temp, sum, index) {
    if (sum > n || temp.length > k) {
        return
    }

    if (sum == n && temp.length == k) {
        result.push([...temp])
        return
    }

    for (let i = index; i <= 9; i++) {
        temp.push(i)
        backtrack(temp, sum + i, i + 1)
        temp.pop()
    }
}

return result

File: combinationsumiii.java
//Java Solution

class Solution {
    public List<List<Integer>> combinationSum3(int k, int n) {
        LinkedList<Integer> temp = new LinkedList<>();
        List<List<Integer>> result = new ArrayList<>();
        int index = 1;
        int sum = 0;
        
        backtrack(temp, result, index, sum, n, k);
        return result;
    }
    
    public void backtrack(LinkedList<Integer> temp, List<List<Integer>> result, int index, int sum, int n, int k) {
        if (sum > n || temp.size() > k) {
            return;
        }
        
        if (sum == n && temp.size() == k) {
            result.add(new ArrayList<>(temp));
            return;
        }
        
        for (int i = index; i <= 9; i++) {
            temp.add(i);
            backtrack(temp, result, i + 1, sum + i, n, k);
            temp.removeLast();
        }
    }
}